### PR TITLE
Patch for device placement (Reduce host and device syncs)

### PR DESCRIPTION
**Pull request by shanmugamr1992**
## What does this PR do?

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Comment by shanmugamr1992** [comment]
The main idea is to reduce host and device sync. 
We also want to use CUDA graphs with DDP. So adding necessary changes.

File: src/lightning/pytorch/strategies/ddp.py [hunk 1]
```diff
@@ -177,7 +177,9 @@ def _setup_model(self, model: Module) -> DistributedDataParallel:
         """Wraps the model into a :class:`~torch.nn.parallel.distributed.DistributedDataParallel` module."""
         device_ids = self.determine_ddp_device_ids()
         log.debug(f"setting up DDP model with device ids: {device_ids}, kwargs: {self._ddp_kwargs}")
-        return DistributedDataParallel(module=model, device_ids=device_ids, **self._ddp_kwargs)
+        with torch.cuda.stream(torch.cuda.Stream()):
```
Comment (carmocca) [inline src/lightning/pytorch/strategies/ddp.py:180]: What's the benefit of this? This is making tests fail

File: src/lightning/pytorch/trainer/connectors/logger_connector/result.py [hunk 2]
```diff
@@ -411,13 +411,13 @@ def register_key(self, key: str, meta: _Metadata, value: _VALUE) -> None:
 
         Value can be provided as a nested collection
         """
-        metric = _ResultMetric(meta, isinstance(value, Tensor)).to(self.device)
```
Comment (carmocca) [inline src/lightning/pytorch/trainer/connectors/logger_connector/result.py:414]: I think this is a good idea. In fact, let's try removing `self.device` completely and let each `_ResultMetric` use the device from each value tensor on registration.

This means something like

```python
# step 1
self.log("x", torch.tensor(1., device="cuda")

# step 2
self.log("x", torch.tensor(2., device="cpu")
```

will now fail, but this might be desirable to avoid silent correctness.

**Review by Borda — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

File: src/lightning/pytorch/strategies/ddp.py [hunk 3]
```diff
@@ -182,7 +183,10 @@ def _setup_model(self, model: Module) -> DistributedDataParallel:
         """Wraps the model into a :class:`~torch.nn.parallel.distributed.DistributedDataParallel` module."""
         device_ids = self.determine_ddp_device_ids()
         log.debug(f"setting up DDP model with device ids: {device_ids}, kwargs: {self._ddp_kwargs}")
-        return DistributedDataParallel(module=model, device_ids=device_ids, **self._ddp_kwargs)
+        # https://pytorch.org/docs/stable/notes/cuda.html#id5
+        ctx = torch.cuda.stream(torch.cuda.Stream()) if torch.cuda.is_available() else nullcontext()
```
Comment (awaelchli) [inline src/lightning/pytorch/strategies/ddp.py:187]: Should we avoid using the stream if not running on CUDA? Even if there are no known side effects, for sanity I would prefer that.

Comment (carmocca) [reply src/lightning/pytorch/strategies/ddp.py:187]: Good point, yes. We can check the root device type instead
